be/src/agent/task_worker_pool.cpp

```diff
@@ -854,76 +750,35 @@ void* TaskWorkerPool::_publish_version_worker_thread_callback(void* arg_this) {
-        batch_publish_latency += publish_latency;
-        g_publish_latency << publish_latency;
-        TFinishTaskRequest finish_task_request;
-        if (!status.ok()) {
-            StarRocksMetrics::instance()->publish_task_failed_total.increment(1);
-            // if publish failed, return failed, FE will ignore this error and
-            // check error tablet ids and FE will also republish this task
-            LOG(WARNING) << "Fail to publish version. signature:" << publish_version_task->signature
-                         << " txn_id: " << publish_version_task->publish_version_req.transaction_id
-                         << " related tablet num: " << tablet_ids.size()
-                         << " error tablet num:" << error_tablet_ids.size() << " time: " << publish_latency << "ms";
-            finish_task_request.__set_error_tablet_ids(error_tablet_ids);
-        } else {
-            LOG(INFO) << "publish_version success. signature:" << publish_version_task->signature
-                      << " txn_id: " << publish_version_task->publish_version_req.transaction_id
-                      << " related tablet num: " << tablet_ids.size() << " time: " << publish_latency << "ms";
-        }
-
-        status.to_thrift(&finish_task_request.task_status);
-        finish_task_request.__set_backend(worker_pool_this->_backend);
-        finish_task_request.__set_task_type(publish_version_task->task_type);
-        finish_task_request.__set_signature(publish_version_task->signature);
-        finish_task_request.__set_report_version(_s_report_version);
-
-        finish_task_requests.emplace_back(std::move(finish_task_request));
+        auto& finish_task = finish_task_requests.emplace_back();
+        finish_task.__set_backend(worker_pool_this->_backend);
+        finish_task.__set_report_version(_s_report_version);
+        run_publish_version_task(*threadpool, publish_version_task, finish_task, affected_dirs);
         priority_tasks.pop();
 
         if (priority_tasks.empty() || finish_task_requests.size() > PUBLISH_VERSION_BATCH_SIZE ||
-            batch_publish_latency > config::max_batch_publish_latency_ms) {
-            // persist all related meta once in a group.
-            tablets.reserve(tablet_ids.size());
-            for (const auto tablet_id : tablet_ids) {
-                TabletSharedPtr tablet = StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id);
-                if (tablet != nullptr) {
-                    tablets.push_back(tablet);
-                }
-            }
-
-            auto st = StorageEngine::instance()->txn_manager()->persist_tablet_related_txns(tablets);
-            if (!st.ok()) {
-                LOG(WARNING) << "failed to persist transactions, tablets num: " << tablets.size() << " err: " << st;
-            }
-
-            tablet_ids.clear();
-            tablets.clear();
-
+            MonotonicMillis() - last_batch_publish_ts > config::max_batch_publish_latency_ms) {
```

Comment:
set last_batch_publish_ts before worker_pool_this->_worker_thread_condition_variable->wait(l), wait task time will be calculated.